src/index.d.ts

```diff
@@ -0,0 +1,1130 @@
+export {};
+
+// This utility is here to handle the case where you have an explicit union
+// between string literals and the generic string type. It would normally
+// resolve out to just the string type, but this generic LiteralUnion maintains
+// the intellisense of the original union.
+//
+// It comes from this issue: microsoft/TypeScript#29729:
+//   https://github.com/microsoft/TypeScript/issues/29729#issuecomment-700527227
+export type LiteralUnion<T extends U, U = string> =
+  | T
+  | (Pick<U, never> & { _?: never | undefined });
+
+export type AST = any;
+export type Doc = doc.builders.Doc;
+
+// The type of elements that make up the given array T.
+type ArrayElement<T> = T extends Array<infer E> ? E : never;
+
+// A union of the properties of the given object that are arrays.
+type ArrayProperties<T> = {
+  [K in keyof T]: NonNullable<T[K]> extends any[] ? K : never;
+}[keyof T];
+
+// A union of the properties of the given array T that can be used to index it.
+// If the array is a tuple, then that's going to be the explicit indices of the
+// array, otherwise it's going to just be number.
+type IndexProperties<T extends { length: number }> = IsTuple<T> extends true
+  ? Exclude<Partial<T>["length"], T["length"]>
+  : number;
+
+// Effectively performing T[P], except that it's telling TypeScript that it's
+// safe to do this for tuples, arrays, or objects.
+type IndexValue<T, P> = T extends any[]
+  ? P extends number
+    ? T[P]
+    : never
+  : P extends keyof T
+  ? T[P]
+  : never;
+
+// Determines if an object T is an array like string[] (in which case this
+// evaluates to false) or a tuple like [string] (in which case this evaluates to
+// true).
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+type IsTuple<T> = T extends []
+  ? true
+  : T extends [infer First, ...infer Remain]
+  ? IsTuple<Remain>
+  : false;
+
+type CallProperties<T> = T extends any[] ? IndexProperties<T> : keyof T;
+type IterProperties<T> = T extends any[]
+  ? IndexProperties<T>
+  : ArrayProperties<T>;
+
+type CallCallback<T, U> = (path: AstPath<T>, index: number, value: any) => U;
+type EachCallback<T> = (
+  path: AstPath<ArrayElement<T>>,
+  index: number,
+  value: any
+) => void;
+type MapCallback<T, U> = (
+  path: AstPath<ArrayElement<T>>,
+  index: number,
+  value: any
+) => U;
+
+// https://github.com/prettier/prettier/blob/next/src/common/ast-path.js
+export class AstPath<T = any> {
+  constructor(value: T);
+
+  get key(): string | null;
+  get index(): number | null;
+  get node(): T;
+  get parent(): T | null;
+  get grandparent(): T | null;
+  get isInArray(): boolean;
+  get siblings(): T[] | null;
+  get next(): T | null;
+  get previous(): T | null;
+  get isFirst(): boolean;
+  get isLast(): boolean;
+  get isRoot(): boolean;
+  get root(): T;
+  get ancestors(): T[];
+
+  stack: T[];
+  callParent<U>(callback: (path: this) => U, count?: number): U;
+  getName(): PropertyKey | null;
+  getValue(): T;
```

Comment:
Can we deprecate these two methods in favor of `.key` / `.index` / `.node` getter?

Reply:
Where is `PropertyKey` ? Can't see.

Reply:
`PropertyKey` is defined in TypeScript (`/node_modules/typescript`)